engine/src/host_imports/imports.rs

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2020 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use super::*;
+use super::ivalues_lifting::wvalues_to_ivalues;
+use super::ivalues_lowering::ivalue_to_wvalues;
+use super::utils::itypes_args_to_wtypes;
+use super::utils::itypes_output_to_wtypes;
+
+use crate::init_wasm_func_once;
+use crate::call_wasm_func;
+use crate::HostImportDescriptor;
+
+use wasmer_core::Func;
+use wasmer_core::vm::Ctx;
+use wasmer_core::typed_func::DynamicFunc;
+use wasmer_core::types::Value as WValue;
+use wasmer_core::types::FuncSig;
+use wasmer_wit::types::RecordType;
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+pub(crate) fn create_host_import_func(
+    descriptor: HostImportDescriptor,
+    record_types: Rc<HashMap<u64, RecordType>>,
+) -> DynamicFunc<'static> {
+    let allocate_func: AllocateFunc = Box::new(RefCell::new(None));
+    let set_result_ptr_func: SetResultPtrFunc = Box::new(RefCell::new(None));
+    let set_result_size_func: SetResultSizeFunc = Box::new(RefCell::new(None));
+
+    let HostImportDescriptor {
+        host_exported_func,
+        argument_types,
+        output_type,
+        error_handler,
+    } = descriptor;
+
+    let output_type_to_types = |output_type| match output_type {
+        Some(ty) => vec![ty],
+        None => vec![],
+    };
+
+    let raw_args = itypes_args_to_wtypes(&argument_types);
+    let raw_output = itypes_output_to_wtypes(&output_type_to_types(output_type));
+
+    let func = move |ctx: &mut Ctx, inputs: &[WValue]| -> Vec<WValue> {
+        init_wasm_func_once!(allocate_func, ctx, i32, i32, ALLOCATE_FUNC_NAME, 2);
+
+        let result = match wvalues_to_ivalues(ctx, inputs, &argument_types, &record_types) {
+            Ok(ivalues) => host_exported_func(ctx, ivalues),
+            Err(e) => {
+                log::error!("error occurred while lifting values in host import: {}", e);
+                error_handler
+                    .as_ref()
+                    .map_or_else(|| default_error_handler(&e), |h| h(&e))
+            }
+        };
+        let wvalues = ivalue_to_wvalues(ctx, result, &allocate_func);
+
+        match wvalues.len() {
```

Comment:
Maybe leave a comment on how 2 is different from 1? Like "2 usually means string or array, while size of 1 usually means a primitive like U8"